core/kernel/console.c

```diff
@@ -128,6 +128,8 @@ void configure_console_from_dt(void)
 	int offs;
 
 	fdt = get_external_dt();
+	if (!fdt)
+		fdt = get_embedded_dt();
```

Comment:
We have `get_dt()` already which does the opposite order (embedded then external). Would it work here? @etienne-lms

Reply:
Yeah, just checked and as you said get_dt() is doing same thing but in opposite manner.
So should i change it to get_dt()

Reply:
Yes I guess it would. But I think the order can be quite important.
Embedded DTB is secure by nature since loaded (and authentified) as part of Core image.
External DTB may be secure or not (actually it must be non-secure since [core_mmu.c](https://github.com/OP-TEE/optee_os/blob/3.8.0/core/arch/arm/mm/core_mmu.c#L671-L673)). Where should it get DTB from first: embedded or external? I think embedded DTB should have precedence.

Also, here we're looking for the console from the chosen node. If we want to find the chosen console from either embedded or external DTB, i think the right sequence should be:
1. find embedded DTB
if found, look up chosen from embedded DTB
if found, look up console data in chosen node
if found use it to configure Core console, we're done.
2. find external DTB
if found, look up chosen from external  DTB
if found, look up console data in chosen node
if found use it to configure Core console, we're done.
3. No chosen console found, we're done.

Reply:
Which case are we trying to cover which would not be properly handled by:
```diff
-	fdt = get_external_dt();
+	fdt = get_dt();
```
?

Reply:
OP-TEE Core console is not strictly speaking a secure feature. Hence I guess a generic way would be to try to get it from either secure or non-secure DTB.
In this scope I think the generic way is to look up for a chosen console in one of the DTBs instead of looking for one of the DTBs and find the chosen console it defines.

However, this remains my understanding of the issue.
@sahilnxp, what is the initial issue you want to solve?

Reply:
That said, if platforms consider external/embedded DTBs are exclusive, then I agree with @jforissier: his proposal addresses the issue.

Reply:
> OP-TEE Core console is not strictly speaking a secure feature

More or less true, but debatable ;-)

> That said, if platforms consider external/embedded DTBs are exclusive

Most of the time I suspect they will be, but even if not the case, I would recommend simplicity here.

> @sahilnxp, what is the initial issue you want to solve?

Very good question :) this is what matters most.

Reply:
The initial issue i want to solve is:
We are enabling the Device Tree for LX2 platform with Embedded DTB option.
But it is always trying to get the External DTB in configure_console_from_dt() function which is called from init_primary_helper()
So if one is using the Embedded way of including in DTB, the current implementation will not work.
Thats why thought of checking the embedded dtb also if external was not found.

With latest commit i was actually trying to implement the sequence which @etienne-lms  mentioned in his first comment.

Yes, in our platform we are considering that either embedded or external dtb will be there.

So I think for simplicity if I agree that we can directly us the `get_dt()` function here.

Reply:
Actually I see a regression in plat-stm32mp1 if replacing here `get_external_dt()` with `get_dt()`.
As a specific console init sequence that explicitly use the secure/embedded DTB. The proposed change will likely break this platform. This should be addressed.

Also: with embedded DTB parsed from this function `configure_console_from_dt()`, you'll get the embedded DTB loaded into the INIT sections of the Core, when `CFG_WITH_PAGER` is enabled. This adds a constraint on internal RAM layout: DTB may be big if based on Linux kernel DTB. This has to be addressed.

Reply:
Ok, I just checked the plat-stm32mp1, they have initialized the Console in `init_console_from_dt()` which is registered as `service_init_late(init_console_from_dt)`.
But since it is registered as late init call, will `init_console_from_dt()` can overwrite the already associated driver and register stm specific driver with that.
No doubt that will be redundant, but is this actually possible ?

Regarding second thing with `CFG_WITH_PAGER`, I actually don't have much idea about this. Can you please help on this how this can be taken care ?

Reply:
> But since it is registered as late init call, will init_console_from_dt() can overwrite the already associated driver and register stm specific driver with that.

Only **plat-stm32mp1** embeds stm32_uart driver, but with some specific resource `configure_console_from_dt()` do not handle. I need to check how to deal with that.

> Regarding second thing with CFG_WITH_PAGER, I actually don't have much idea about this. Can you please help on this how this can be taken care ?

Sure I'll try to help here. This is quite specific to the platforms that enable pager support. And each of these platforms (there ain't many) have their specific constraints.

I need to investigate a bit.

Reply:
I think I have something that should satisfy everybody. It is still drafty but I think our dear maintainers will help us to get it into a clean state. Please see https://github.com/OP-TEE/optee_os/pull/3834.

Reply:
FYI, contrary to my comment https://github.com/OP-TEE/optee_os/pull/3827#discussion_r420148075, there no issue with stm32mp1 console specific inits since it does not register the the `dt_drivers` framework, hence is not probed by `init_console_from_dt()`.

 #3834 should solve the 2nd issue I mentioned about pager _init_ sections constraints.